### PR TITLE
 Refactor dependency injection - cluster and env 

### PR DESCRIPTION
**Pull request by fmoral2**
#### Proposed Changes ####

- https://github.com/rancher/qa-tasks/issues/1069 

- This has many files changed cause it needed to change due to dependency, but whats changed in fact is very small changes:
- Env and Cluster dependencies

- Refactor how we inject dependencies in a way of create a more standardized and clear model which basically is:

    . Send dependencies top down injecting in the main and passing where needed

  Benefits:

   . Have a better visibility and understanding of your deps.
   . Have a easier way to follow through to debug and maintain
   . Standardize
   . Have deps available without need to duplicate code
   . Simplify process of adding dependencies ( now you just need to add and pass along in the main)

   

#### Testing ####

Checklist:
1. If your PR changes anything on or related to Jenkins, run it pointing to your branch to make sure it's okay.

- yes 
2. Verify code lint; we should not have errors.

- yes 

3. Update the documentation if needed.

- will open a PR only for README 
4. Update makefile and docker run if adding new tests.
- n/a

5. Run your tests at least 4 times with all configurations needed and possible.
- yes 

6. If needed test with different os types.

- yes

### CODE REVIEW

**Comment by fmoral2** [comment]
https://github.com/rancher/qa-tasks/issues/1069

**Comment by rancher-max** [comment]
Needs rebase on main again!

**Review by rancher-max — COMMENTED** [review]
I think this looks good after rebase. We might not want to delete all the versionbump tests here -- I know the versionbump PR updated those so perhaps after rebase that'll be resolved, otherwise maybe keep canal and cilium tests there?

**Comment by fmoral2** [comment]
> I think this looks good after rebase. We might not want to delete all the versionbump tests here -- I know the versionbump PR updated those so perhaps after rebase that'll be resolved, otherwise maybe keep canal and cilium tests there?

it will all be solved after rebase yep

**Comment by fmoral2** [comment]
> Needs rebase on main again!

i will wait all PRs get merged, so i dont complicate people lives with this updates.

**Review by VestigeJ — DISMISSED** [review]

File: config/reader.go
Comment (mdrahman-suse) [inline config/reader.go:18]: can we rename it to `logger`

Comment (fmoral2) [reply config/reader.go:18]: here we cant, its the pkg name

i can rename to `log`, if thats better

File: entrypoint/createcluster/createcluster_suite_test.go
Comment (mdrahman-suse) [inline entrypoint/createcluster/createcluster_suite_test.go:23]: Any reason why we are adding a variable here and replacing the one liner?

File: entrypoint/deployrancher/rancher_suite_test.go
Comment (mdrahman-suse) [inline entrypoint/deployrancher/rancher_suite_test.go:21]: Nit: extra spacing?

Comment (fmoral2) [reply entrypoint/deployrancher/rancher_suite_test.go:21]: why ?  they are on the same level 

```
var (
	cluster *factory.Cluster
	flags   *customflag.FlagConfig
)
```

File: entrypoint/createcluster/createcluster_suite_test.go
Comment (fmoral2) [reply entrypoint/createcluster/createcluster_suite_test.go:23]: yeah its just a cleaner way to not tie this error directly , like first get whats happening then you exit
so if you need to deal with some extra test stuff, like log ,clean something or whatever you have a way to do so +

right now, makes no diff logically, just make more readable

Comment (rancher-max) [reply entrypoint/createcluster/createcluster_suite_test.go:23]: I was going to comment the same thing 😅 I prefer the one liner myself, but I'm fine either way here.

Comment (fmoral2) [reply entrypoint/createcluster/createcluster_suite_test.go:23]: well let me know....this PR its really getting stuck here

Comment (fmoral2) [reply entrypoint/createcluster/createcluster_suite_test.go:23]: here on the PRs, not on this comment i meant

Comment (VestigeJ) [reply entrypoint/createcluster/createcluster_suite_test.go:23]: I also think we should keep the one-liner os.Exit(m.Run())

File: factory/cluster.go
Comment (mdrahman-suse) [inline factory/cluster.go:17]: Rename to log

File: pkg/assert/node.go
Comment (mdrahman-suse) [inline pkg/assert/node.go:31]: typo `\ninfo`?

File: pkg/assert/pod.go
Comment (mdrahman-suse) [inline pkg/assert/pod.go:105]: can we store and fetch KubeConfigFile from cluster var?

File: entrypoint/createcluster/createcluster_suite_test.go
Comment (fmoral2) [reply entrypoint/createcluster/createcluster_suite_test.go:23]: sure, removing then

File: factory/config.go
Comment (fmoral2) [reply factory/config.go:33]: i think it was my bad on refactor other stuff naming , i really agree also. fixed tks

File: pkg/assert/node.go
Comment (fmoral2) [reply pkg/assert/node.go:31]: nope, its to break line before the log

File: pkg/assert/pod.go
Comment (fmoral2) [reply pkg/assert/pod.go:105]: this one as per how its used now, i would say lets keep global

File: pkg/assert/node.go
Comment (mdrahman-suse) [reply pkg/assert/node.go]: should probably use Enums/Constants for represent the log level

Comment (fmoral2) [reply pkg/assert/node.go]: could be a possible refactor in future

File: factory/cluster.go
Comment (rancher-max) [inline factory/cluster.go:107]: possible to use BasePath function here at all?

**Review by rancher-max — DISMISSED** [review]
Most of the stuff I have is nitpicky, so I've approved unless you want to talk about any of it or change any of it based on the comments.

Good job on this! Lots of refactoring for it all.

File: entrypoint/upgradecluster/upgradesuc_test.go
Comment (fmoral2) [reply entrypoint/upgradecluster/upgradesuc_test.go:82]: no, lets remove!  tks

File: shared/cluster.go
Comment (fmoral2) [reply shared/cluster.go:57]: oh please no ! lol 

i added like this to prevent adding a "random" param on the signature, keeping this lighter , since this function its called a lot of times everywhere 
ans this is not a optional var today, so we are always going to have this env arch

File: factory/config.go
Comment (fmoral2) [reply factory/config.go:64]: today we cant cause of circular dependency between them

```
# github.com/rancher/distros-test-framework/entrypoint/createcluster
package github.com/rancher/distros-test-framework/entrypoint/createcluster (test)
        imports github.com/rancher/distros-test-framework/factory
        imports github.com/rancher/distros-test-framework/shared
        imports github.com/rancher/distros-test-framework/factory: import cycle not allowed
```

File: factory/cluster.go
Comment (fmoral2) [reply factory/cluster.go:107]: https://github.com/rancher/distros-test-framework/pull/79#discussion_r1638123337

**Review by rancher-max — APPROVED** [review]

**Comment by mdrahman-suse** [comment]
I see the use of this `cluster` variable in almost all the the tests so I was wondering if it makes sense to have it initialized as a global var during the test lifecycle instead of needing to pass it along every time we write the test. Sharing my thought and we can discuss if needed...

**Comment by fmoral2** [comment]
> I see the use of this `cluster` variable in almost all the the tests and suite so I was wondering if it makes sense to have it initialized as a global var during the test lifecycle instead of needing to pass it along every time we write the test. Sharing my thought and we can discuss if needed...


 what are the gains you see initializing  it somewhere else and injecting here as global?

**Comment by mdrahman-suse** [comment]
> > I see the use of this `cluster` variable in almost all the the tests and suite so I was wondering if it makes sense to have it initialized as a global var during the test lifecycle instead of needing to pass it along every time we write the test. Sharing my thought and we can discuss if needed...
> 
> what are the gains you see initializing it somewhere else and injecting here as global?

The gain from my perspective is not to provide that var each time we are defining the tests as we have to use that var in the test anyways. This thought came to my mind as I have used DI with Java before and remember to simply inject the dependency (as an interface) on the test file itself (for us it is similar to testcase pkg).

**Comment by fmoral2** [comment]
> > > I see the use of this `cluster` variable in almost all the the tests and suite so I was wondering if it makes sense to have it initialized as a global var during the test lifecycle instead of needing to pass it along every time we write the test. Sharing my thought and we can discuss if needed...
> > 
> > 
> > what are the gains you see initializing it somewhere else and injecting here as global?
> 
> The gain from my perspective is not to provide that var each time we are defining the tests as we have to use that var in the test anyways. This thought came to my mind as I have used DI with Java before and remember to simply inject the dependency (as an interface) on the test file itself (for us it is similar to testcase pkg). Sharing an example here with Java https://www.vogella.com/tutorials/DependencyInjection/article.html Again this is just as I understand DI

so we dont have that kind of stuff in go because one of the ideas is exactly do the opposite of that like prefering using more explicitness and simplicity, like creating your stuff , handling yourself and passing it through ...making things easier to track .
```
New( my interface for some stuff  ) *stuff 
{. return the interface of my stuff } 
```
  those kind of annotations from that link also for me just bring more complexity cause it can hide potential data or stuff from your obecjt

**Comment by mdrahman-suse** [comment]
> > > > I see the use of this `cluster` variable in almost all the the tests and suite so I was wondering if it makes sense to have it initialized as a global var during the test lifecycle instead of needing to pass it along every time we write the test. Sharing my thought and we can discuss if needed...
> > > 
> > > 
> > > what are the gains you see initializing it somewhere else and injecting here as global?
> > 
> > 
> > The gain from my perspective is not to provide that var each time we are defining the tests as we have to use that var in the test anyways. This thought came to my mind as I have used DI with Java before and remember to simply inject the dependency (as an interface) on the test file itself (for us it is similar to testcase pkg). Sharing an example here with Java https://www.vogella.com/tutorials/DependencyInjection/article.html Again this is just as I understand DI
> 
> so we dont have that kind of stuff in go because one of the ideas is exactly do the opposite of that like prefering using more explicitness and simplicity, like creating your stuff , handling yourself and passing it through ...making things easier to track .
> 
> ```
> New( my interface for some stuff  ) *stuff 
> {. return the interface of my stuff } 
> ```
> 
> those kind of annotations from that link also for me just bring more complexity cause it can hide potential data or stuff from your obecjt

Gotcha! Thanks for clarifying. Looks great overall 👍

**Review by mdrahman-suse — APPROVED** [review]
LGTM as long as it has been rebased with the recent changed and ran on Jenkins.

**Comment by mdrahman-suse** [comment]
I think we should run the batch test on this branch before we merge, WDYT?

**Comment by rancher-max** [comment]
Good idea -- I can do that when I run the tests on the RCs tomorrow!

**Comment by fmoral2** [comment]
> I think we should run the batch test on this branch before we merge, WDYT? CC. @rancher-max @fmoral2

good idea

**Review by rancher-max — DISMISSED** [review]

**Review by mdrahman-suse — APPROVED** [review]
LGTM

**Review by rancher-max — APPROVED** [review]

**Pull request merged**
